Interface to build a monomer-based representation of a molecule

rachelnwalker opened the pull request:
This PR includes the following:

- Interface to build a “MonomerMol” (in MonomerMol.cpp)
- `toMonomeric` -- converts an atomistic structure with PDB residue information into a "MonomerMol"
- `toAtomistic.` -- converts a "MonomerMol" into an atomistic `RWMol` (without coordinates)
- `MonomerDatabase.cpp` -- interface for Monomer DB interaction, currently uses maps to store monomer information to simplify this PR
- `testMonomerMol.cpp` -- shows use of functionality introduced in this PR

## What isn’t included in the PR

There are several things we’ve implemented that aren’t included in this PR that we could port:

- Monomer DB support (using sqlite3, with a DB schema similar to https://github.com/PistoiaHELM/HELMMonomerSets/blob/master/HELMCoreLibrary/HELMCoreLibrary.json)
- SMILES monomers — used when monomer is not identified in monomer DB
- HELM reader and writer that uses “MonomerMol”
- FASTA reader and writer that uses “MonomerMol”
- Coordinate and image generation for “MonomerMol” (see slides from 2024 RDKit UGM)
- Registration
- Residue identification via SMARTS matching

## About MonomerMol

Currently, `MonomerMol` isn’t a new data structure — its an RDKit mol with specific properties set when built using the functions in `MonomerMol.h`. We decided to use an RDKit mol instead of making a new data structure so that we could easily use existing APIs such as substructure search, fingerprinting, and serialization as well as our sketcher.

The design of “MonomerMol” is largely inspired by HELM https://pubs.acs.org/doi/10.1021/ci3001925. `AtomPDBResidueInfo` is used to store the monomer ID as `residueName` polymer type and chain ID as `chainId` on the atoms. That information together can be used as a key to retrieve monomer definitions, currently in the form of SMILES, from a monomer database (i.e., C + PEPTIDE, C + RNA, or C + CHEM are all different monomers). 

The monomer definitions include attachment points that are used to identify linkage types between monomers. Bonds on a “MonomerMol” define have a string property `LINKAGE` set to RX-RY , where X is the attachment point from the beginAtom() monomer and Y is the attachment point in the endAtom() monomer. For example, a backbone bond connecting two cysteines would have `LINKAGE` set to `R2-R1`, where a disulfide bond would be set to `R3-R3`. For reference, this is how cysteine is stored in the monomer DB:

<img width="862" alt="image" src="https://github.com/user-attachments/assets/2adf75b0-06fb-4add-b335-3f9293f11e04" />

We’re hoping there is a nice way to incorporate this into existing related RDKit tools such as abbreviations and the recent SCSR parser.

rachelnwalker committed:
(WIP) MonomerMol representation

rachelnwalker committed:
get building

rachelnwalker committed:
update tests

rachelnwalker committed:
Merge remote-tracking branch 'origin/master' into monomer_mol_wip

rachelnwalker committed:
clean up

rachelnwalker committed:
Merge remote-tracking branch 'origin/master' into monomer_mol_wip

rachelnwalker commented:
Still working on fixing the tests!

greglandrum commented:
> Still working on fixing the tests!

Let me know when you want me to take a look

whosayn commented on Code/GraphMol/MonomerMol/AtomisticToMM.cpp:
```diff
@@ -0,0 +1,216 @@
+// symbols that are already in the monomer database. We may need to
+// figure out how to have multiple 3-letter codes for a single symbol
+// polymer_type pair (Histidine is the best example)
+const std::unordered_map<std::string, char> backup_res_table({
```
still little confused by why we need this after reading the docs. is this in place of a user-provided db?

rachelnwalker replied:
Yeah I will fix the doc here this is not well phrased.

Currently, monomer types and HELM symbol are the primary key to the monomer database and only 1 PDB code is allowed for each entry. This means we can't store multiple pronated states of monomers (for example, histidine) in the DB. This map allows commonly seen alternate 3-letter codes to be translated into HELM symbols

whosayn replied:
ooh, I think this rings a bell. This you consider `std::multimap` or `std::unordered_multimap`? I feel like the snippet in https://en.cppreference.com/w/cpp/container/multimap/equal_range resembles your intended use case

whosayn commented on Code/GraphMol/MonomerMol/AtomisticToMM.cpp:
```diff
@@ -0,0 +1,216 @@
+{
+    // Make RWMol and remove waters
+    RDKit::RWMol mol(input_mol);
+    remove_waters(mol);
```
why do we remove the waters?

rachelnwalker replied:
Otherwise they will be turned into monomers on the `MonomerMol`; I guess we could ignore them in the conversion but we are working on a copy anyway so this seems simpler

Technically, I think we want to remove all solvents

whosayn replied:
make sense. I'd hope that behavior would be opt-in since this would mean that the "compression" wouldn't be lossless by default.

whosayn commented on Code/GraphMol/MonomerMol/MMToAtomistic.cpp:
```diff
@@ -0,0 +1,319 @@
+        res_info->setResidueNumber(residue_number);
+        res_info->setResidueName(residue_name);
+        // to be consistent with the rdkit adapter and RDKit's own PDB writer
+        res_info->setInsertionCode(" ");
```
our rdkit adapter?

rachelnwalker replied:
oops, I'll remove that

whosayn commented on Code/GraphMol/MonomerMol/MMToAtomistic.cpp:
```diff
@@ -0,0 +1,319 @@
+        res_info->setInsertionCode(" ");
+
+        std::string pdb_name;
+        if (atom->getPropIfPresent(ATOM_PDB_NAME_PROP, pdb_name)) {
```
Should the atom pdb name be the same as the `_Name` property for residue atoms?

rachelnwalker replied:
I don't think I understand your question;`ATOM_PDB_NAME_PROP` is set on atoms in the monomer DB. 

Is `_Name` usually used to store atom PDB names in RDKit?

whosayn replied:
oh, it was a question. I just find it odd that the atom would have two "names". `Is _Name usually used to store atom PDB names in RDKit?` I don't think so, but if we were to do that II don't think it'd be a bad idea

whosayn commented on Code/GraphMol/MonomerMol/MMToAtomistic.cpp:
```diff
@@ -0,0 +1,319 @@
+    }
+}
+
+ChainType get_chain_type(std::string_view polymer_id)
```
we should remove this

rachelnwalker replied:
its currently used below -- maybe rename to `chain_name_to_type`?

whosayn replied:
my bad :)

whosayn commented on Code/GraphMol/MonomerMol/MonomerMol.cpp:
```diff
@@ -0,0 +1,339 @@
+        }
+    }
+
+    std::string annotation{};
```
we should probably remove this for now

whosayn commented on Code/GraphMol/MonomerMol/MonomerMol.h:
```diff
@@ -0,0 +1,109 @@
+#include <string>
+#include <string_view>
+
+const std::string ANNOTATION{"ANNOTATION"};
```
I feel like most of these should be removed for now

whosayn commented on Code/GraphMol/MonomerMol/MonomerMol.cpp:
```diff
@@ -0,0 +1,339 @@
+#include <boost/functional/hash.hpp>
+
+template <typename T>
+static bool is_dummy_obj(const T& obj)
```
saw a compiler error about this. probably remove `obj` or mark it as `[[maybe_unused]]`

whosayn commented on Code/GraphMol/MonomerMol/MonomerDatabase.h:
```diff
@@ -0,0 +1,32 @@
+{
+
+enum class ChainType;
+class [[nodiscard]] MonomerDatabase : public boost::noncopyable
```
we should probably remove references to `HELM`

whosayn commented on Code/GraphMol/MonomerMol/MonomerMol.h:
```diff
@@ -0,0 +1,109 @@
+#include <RDGeneral/export.h>
+
+#include <string>
+#include <string_view>
```
looks like we're missing `#include <vector>`

whosayn commented on Code/GraphMol/MonomerMol/MonomerMol.cpp:
```diff
@@ -0,0 +1,339 @@
+    a->setProp(ATOM_LABEL, n); // temporary, for image generation
+    // Always start with BRANCH_MONOMER as false, will be set to
+    // true if branch linkage is made to this monomer
+    a->setProp(BRANCH_MONOMER, false);
```
do we need these?

rachelnwalker replied:
No; I left them to make merging the HELM parser/coordinate generation/drawing code easier in the future, but it seems to just make this more confusing for now so I'll remove it

whosayn replied:
ahh, I see. I feel like without the context of our image gen and HELM stuff it's a bit confusing

whosayn commented on Code/GraphMol/MonomerMol/Conversions.h:
```diff
@@ -0,0 +1,38 @@
+ * @return Atomistic molecule
+ */
+RDKIT_MONOMERMOL_EXPORT boost::shared_ptr<RDKit::RWMol>
+monomerMolToAtomsitic(const RDKit::ROMol& monomer_mol);
```
typo

whosayn reviewed (commented):
Added some comments about removing some leftover helm stuff and the build errors. The tests looks good, but I'll look closer after it starts passing

rachelnwalker committed:
husseins comments + fix build

whosayn commented on Code/GraphMol/MonomerMol/MonomerDatabase.cpp:
```diff
@@ -0,0 +1,174 @@
+}
+
+[[nodiscard]] MonomerDatabase::monomer_smiles_t
+MonomerDatabase::get_monomer_smiles(std::string monomer_id,
```
looks like the `monomer_type` parameter needs [[maybe_unused]]

rachelnwalker committed:
Merge branch 'rdkit:master' into monomer_mol_wip

rachelnwalker committed:
added [[maybe_unused]]

rachelnwalker commented:
@greglandrum this is ready for you to take a look at!

greglandrum reviewed (commented):
I didn't spend a lot of time deeply in the code here; I mainly looked at the headers and the API.
I think the outline of things looks solid and I can see how this could be quite useful. There are a bunch of "style guide" compliance things, but that should be mostly mechanical.

We should think a bit to make sure that the design is flexible enough to allow us to use it for more general cases like polymers. I suspect it's ok, but the way the linkages are handled may need to be tweaked.

rachelnwalker committed:
Merge branch 'rdkit:master' into monomer_mol_wip

rachelnwalker committed:
checkpoint

rachelnwalker committed:
updated code

rachelnwalker committed:
get tests working

rachelnwalker committed:
Merge remote-tracking branch 'origin/master' into monomer_mol_wip

rachelnwalker committed:
Merge branch 'monomer-mol-pr2' into monomer_mol_wip

rachelnwalker committed:
Merge remote-tracking branch 'origin/master' into monomer_mol_wip

greglandrum commented on Code/GraphMol/MonomerMol/MonomerMol.h:
```diff
@@ -0,0 +1,127 @@
+    // std::string polymer_id;
+};
+
+enum class ChainType { PEPTIDE, RNA, DNA, CHEM };
```
Maybe these enums could be `BETTER_ENUM`s so that you get the enum<->string conversion automatically?

There are examples of these being used in `RGroupDecompParams`, `CXSmilesFields`, etc.

greglandrum commented on Code/GraphMol/MonomerMol/Conversions.h:
```diff
@@ -0,0 +1,38 @@
+ * @return Atomistic molecule
+ */
+RDKIT_MONOMERMOL_EXPORT boost::shared_ptr<RDKit::RWMol>
+monomerMolToAtomsitic(const RDKit::ROMol& monomer_mol);
```
```suggestion
monomerMolToAtomistic(const RDKit::ROMol& monomer_mol);
```

greglandrum commented on Code/GraphMol/MonomerMol/AtomisticToMM.cpp:
```diff
@@ -0,0 +1,874 @@
+#include "Conversions.h"
```
please add copyright to all source files

greglandrum commented on Code/GraphMol/MonomerMol/Conversions.h:
```diff
@@ -0,0 +1,38 @@
+ * monomer boundaries
+ * @return MonomerMol
+ */
+RDKIT_MONOMERMOL_EXPORT boost::shared_ptr<RDKit::RWMol>
```
Can't we return `std::unique_ptr` from here? Or, at the least, `std::shared_ptr`?

greglandrum commented on Code/GraphMol/MonomerMol/Conversions.h:
```diff
@@ -0,0 +1,38 @@
+ * @param monomer_mol Monomeristic molecule to convert to atomistic
+ * @return Atomistic molecule
+ */
+RDKIT_MONOMERMOL_EXPORT boost::shared_ptr<RDKit::RWMol>
```
same question about the return type

greglandrum commented on Code/GraphMol/MonomerMol/testMonomerMol.cpp:
```diff
@@ -0,0 +1,171 @@
+    addConnection(monomer_mol, midx3, midx4, ConnectionType::FORWARD);
+    addConnection(monomer_mol, midx4, midx5, ConnectionType::FORWARD);
+
+    CHECK(std::string(">Chain A\nRD\n>Chain B\nKIT") == to_fasta(monomer_mol));
```
This is a good integration-type test, but I think it would also be useful to add a check of the number of atoms and bonds in `monomer_mol`

greglandrum commented on Code/GraphMol/MonomerMol/testMonomerMol.cpp:
```diff
@@ -0,0 +1,171 @@
+    CHECK(atomistic_mol);
+
+    // PDB info is used to convert the atomistic sturcture into a MonomerMol
+    auto monomer_mol = atomisticToMonomerMol(*atomistic_mol);
```
please check the number of atoms and bonds here too

greglandrum reviewed (commented):
Just reviewed the tests and the headers and I think the idea is useful.
I made some comments on what I've looked at so far.

Reviewing the code will take longer, so I'd like to hold off on that until after we're all happy with the API

rachelnwalker committed:
Merge branch 'rdkit:master' into monomer_mol_wip

rachelnwalker committed:
A few changes

* Removed SMARTS matching method for atomistic -> monomeric, that will be
    in separate PR
* Addressed review comments
* Changed some naming schemes
* Added tests

rachelnwalker committed:
fix mac build

rachelnwalker commented:
Hi @greglandrum -- this is ready for you to take a look at!

I removed a chunk of code that was dedicated to automatic residue identification because I think that would be better suited for a separate PR, updated some naming based on suggestions from @cdvonbargen, addressed your comments, and added several tests.

I am thinking about updating the `FORWARD` and `SIDECHAIN` linkages to be dative bonds so that there is a sense of directionality than can be considered in operations such as substructure matching.

rachelnwalker committed:
Merge branch 'rdkit:master' into monomer_mol_wip

rachelnwalker committed:
Merge branch 'rdkit:master' into monomer_mol_wip
